File: infra/src/main/java/mixter/infra/repositories/AggregateRepository.java
package mixter.infra.repositories;

import mixter.domain.AggregateId;
import mixter.domain.Event;
import mixter.infra.EventStore;

import java.util.List;
import java.util.NoSuchElementException;

public abstract class AggregateRepository<T> {
    protected EventStore eventStore;

    public AggregateRepository(EventStore eventStore) {
        this.eventStore = eventStore;
    }

    public T getById(AggregateId aggregateId) {
        List<Event> history = eventStore.getEventsOfAggregate(aggregateId);
        if (history.isEmpty()) {
            throw new NoSuchElementException();
        } else {
            return fromHistory(history);
        }
    }

    protected abstract T fromHistory(List<Event> history);
}

File: infra/src/main/scala/mixter/infra/repositories/AggregateRepository.scala
package mixter.infra.repositories

import mixter.domain.Aggregate
import mixter.infra.EventStore

abstract class AggregateRepository[A<:Aggregate]{
  protected def store:EventStore
  def getById(id: A#Id):Option[A] ={
    val history = store.eventsOfAggregate(id)
    history.headOption.map( initialEvent=>
      build(initialEvent.asInstanceOf[A#InitialEvent], history.tail.map(_.asInstanceOf[A#AggregateEvent]))
    )
  }
  protected def build(initialEvent:A#InitialEvent, history:Seq[A#AggregateEvent]):A
}
